### ISSUE TITLE
Support for Neovim's built-in LSP client

### ISSUE THREAD

**Issue by kmhofmann** [jackguo380/vim-lsp-cxx-highlight#17]
Recently, Neovim added native LSP support by means of a built-in client. See https://neovim.io/doc/user/lsp.html and https://github.com/neovim/nvim-lsp for a configuration helper repo.

This obviates the need to use any of the supported LSP client plugins of vim-lsp-cxx-highlight, and hence I am using neither. The built-in client works great for me in combination with *ccls*.

Is support for the Neovim-native LSP client on the roadmap?

**Comment by jackguo380** [comment]
Hi kmhofmann,

Thanks for the suggestion, I took a look and found it was easier to implement than I thought so I just pushed support for it.

Tested it locally on `nvim v0.5.0-404-g49cd750d6` (appimage) and it looks to be working well indexing ccls' codebase.

Please test it out and let me know if you run into any issues. Make sure to include the extra settings I recommend [here](https://github.com/jackguo380/vim-lsp-cxx-highlight/blob/master/sample-configs/nvim-lsp-register.vim)

**Comment by kmhofmann** [comment]
Hi jackguo380 - that was faster than light, and it also works for me.
Thanks a bunch!

**Comment by kmhofmann** [comment]
A possible regression: When trying to use this plugin in combination with an installed LanguageClient_neovim, on an nvim-lsp capable system, semantic highlighting does not work. Could this be connected to it trying to register both?

    $ tail -f /tmp/vim-lsp-cxx-hl.log
    Sat Feb 29 11:14:16 2020: vim-lsp not detected
    Sat Feb 29 11:14:16 2020: LanguageClient#registerHandlers() ret = 0
    Sat Feb 29 11:14:16 2020: LanguageClient-neovim successfully registered
    Sat Feb 29 11:14:16 2020: coc.nvim not detected
    Sat Feb 29 11:50:27 2020: lsp_cxx_hl beginning initialization...
    Sat Feb 29 11:50:27 2020: vim-lsp not detected
    Sat Feb 29 11:50:27 2020: LanguageClient#registerHandlers() ret = 0
    Sat Feb 29 11:50:27 2020: LanguageClient-neovim successfully registered
    Sat Feb 29 11:50:27 2020: coc.nvim not detected
    Sat Feb 29 11:50:27 2020: nvim-lsp successfully registered

**Comment by jackguo380** [comment]
I'm not seeing the issue. I just tried out both LanguageClient (which I normally use) together with nvim-lsp with both of them registering ccls as a language server, in this case the highlighting still works fine.

Its probably something about your configuration that is causing the problem. A potential issue is that you have ccls configured correctly in 1 client but not another then the "bad" ccls is overwriting the highlight when it cannot properly parse the code. Could you provide the relevant bits of your `init.vim` or try with something along the lines of a minimal vimrc?

**Comment by kmhofmann** [comment]
I don't have an issue, as long as both `neovim/nvim-lsp` and `autozimu/LanguageClient-neovim` are actually installed, e.g. via vim-plug. In my test case, `neovim/nvim-lsp` is not cloned, if `autozimu/LanguageClient-neovim` is.

See https://github.com/kmhofmann/dotfiles/blob/a42322065b55f0b3f66264ebe6c2c12e1e5177e3/init.vim#L221. where selection of the plugin to be clones takes place, and https://github.com/kmhofmann/dotfiles/blob/a42322065b55f0b3f66264ebe6c2c12e1e5177e3/init.vim#L936 as well as https://github.com/kmhofmann/dotfiles/blob/a42322065b55f0b3f66264ebe6c2c12e1e5177e3/init.vim#L1002 where the two different clients are configured.

If I remove the if/else clauses in line 221 (link above) and allow *both* cases, then highlighting works without modifications to your plugin. But that's not what a user might want. Think of the case of an LSP-native capable Neovim (i.e. `has('nvim-0.5')`, but where a user wants to use `autozimu/LanguageClient-neovim` only.

**Comment by jackguo380** [comment]
I don't think there's any issue with multiple plugins being available. I tested these cases on my own configuration with both plugins enabled:
| ccls registered in LanguageClient | ccls registered in nvim-lsp | Highlighting works |
| ------------------ | ---------------- | ---------------|
| No | No | No (of course)|
| Yes | No | Yes |
| No | Yes | Yes |
| Yes | Yes | Yes|

These cases still work if I only have either LanguageClient or nvim-lsp installed (Not both).

I think you may have a configuration issue with LanguageClient and the reason enabling nvim-lsp fixes it is because vim-lsp-cxx-highlight is simply receiving the highlighting information from there instead of LanguageClient. I noticed that you don't set `g:LanguageClient_rootMarkers` which I have as:
```vim
let g:LanguageClient_rootMarkers['c'] = ['compile_commands.json', '.ccls_root']
let g:LanguageClient_rootMarkers['cpp'] = ['compile_commands.json', '.ccls_root']
```
Are you certain that LanguageClient is correctly detecting your project's root and that there is a `.ccls` or `compile_commands.json` file there?

**Comment by kmhofmann** [comment]
> Are you certain that LanguageClient is correctly detecting your project's root and that there is a .ccls or compile_commands.json file there?

Yes, LanguageClient_neovim prints the project root correctly upon loading of a file, and there is a `compile_commands.json` file there. I see all code diagnostics emitted by LanguageClient_neovim.

I'll try to reduce my init.vim to a minimal file that can either demonstrate the issue, or where a bug may become apparent.

**Comment by jackguo380** [comment]
Hm that's really strange then, perhaps try enabling `let g:lsp_cxx_hl_verbose_log = 1` and posting the log here.

**Comment by kmhofmann** [comment]
Consider [this](https://gist.github.com/kmhofmann/3ddca1c7ee4276f399ca17d951aee2dc) pretty self-contained init.vim.

Scenario 1 does not lead to semantic highlighting, using the `20f0f7c` commit you made today.
On the other hand, in Scenario 4, semantic highlighting does work for me. The only difference is that I use the commit prior (`71490fd`).

In both cases, I am attempting to use only LanguageClient-neovim, in combination with a Neovim version that supports native LSP (`NVIM v0.5.0-404-g49cd750d6`).

When switching to different scenarios, I always did a `rm -rf autoload/ plugged/` in `~/.config/nvim`, followed by a `:PlugInstall` inside nvim.

Can you reproduce this behavior?

**Comment by Shatur95** [comment]
Hello! Thank you for adding support for nvim-lsp. I switched from `clang` to `ccls`, tried your sample config and the highlighting works. But auto-completion with `nvim-lsp` and `ccls` not works as expected. Do you have the same issue?
[![asciicast](https://asciinema.org/a/pClJKER6R00X2dbWY2XUSH0on.svg)](https://asciinema.org/a/pClJKER6R00X2dbWY2XUSH0on)

**Comment by jackguo380** [comment]
@kmhofmann I tried out the minimal init.vim and I am seeing that semantic highlighting works in all 4 cases. I was worried my personal computer may affect it so I used a competely fresh Ubuntu 18.04 AWS VM I had.

Here are my exact steps to test it. I use my usual test case of using ccls to index its own code base:
```sh
#!/bin/bash
set -e

# Download neovim nightlly (NVIM v0.5.0-404-g49cd750d6)
wget https://github.com/neovim/neovim/releases/download/nightly/nvim.appimage
chmod +x nvim.appimage

# Download minimal init.vim
wget https://gist.github.com/kmhofmann/3ddca1c7ee4276f399ca17d951aee2dc/raw/2f3758c7bd28d521225598b7736bd7556668d6ac/init.vim

# Install plugins
./nvim.appimage -u init.vim +PlugInstall +qall

# Build ccls
git clone --depth 1 https://github.com/MaskRay/ccls
git submodule update --init
# Change this as you need to install build dependencies on your platform
sudo apt install cmake g++ llvm-9 llvm-9-dev clang-9 libclang-9-dev
mkdir build
cd build
cmake .. -DCMAKE_PREFIX_PATH=$(llvm-config-9 --prefix) -DCMAKE_EXPORT_COMPILE_COMMANDS=1
make ccls
cd ..

# Ensure ccls is on PATH
export PATH="$PWD/build:$PATH"

# Get compile commands for ccls
cp build/compile_commands.json .

# View a file
# (Note that it takes a few minutes before semantic highlighting info is available on first index)
../nvim.appimage -u ../init.vim src/indexer.cc
```
After it indexed I see that semantic highlighting pops up as expected.

Could you post your logs for LanguageClient and vim-lsp-cxx-highlight?
```vim
let g:lsp_cxx_hl_log_file = '/tmp/lsp-cxx-hl.log'
let g:lsp_cxx_hl_verbose_log = 1

let g:LanguageClient_loggingFile = '/tmp/languageclient.log'
```

**Comment by jackguo380** [comment]
Hi @Shatur95 

I also use mucomplete, however I use it with LanguageClient-neovim and haven't seen major issues with it. Occasionally mucomplete inserts a bit of extra text like you show in your video however for the most part the completion is working correctly for me.

But of course this has nothing to do with vim-lsp-cxx-highlight. I recommend you ask for support from mucomplete if you haven't done so already.

**Comment by Shatur95** [comment]
Yes, I'm talking about extra text. I'm sorry, I know this is not related to `vim-lsp-cxx-highlight`. Just trying to understand how other people configure `ccls` with `nvim-lsp`. And I'm not sure if this is a MUcomplete or nvim-lsp problem :(

**Comment by jackguo380** [comment]
I missed this part earlier but you said that you swapped from `clangd` to `ccls` and this issue started happening. Its also possible that `ccls` is the cause.

**Comment by Shatur95** [comment]
Tested it via calling `omnifunc` directly, but have the same issue. So, I think this is not `MUcomplete` issue. And since you don't have this issue with `LanguageClient-neovim`, I think that is not a `ccls` problem. So, will report it to the `nvim-lsp`. Thank you!

**Comment by kmhofmann** [comment]
Hi @jackguo380 , can you please try the following?
Instead of calling Neovim in your example above with `../nvim.appimage -u ../init.vim src/indexer.cc`, can you call it simply with `../nvim.appimage -u ../init.vim` and open the C++ file with `:e src/indexer.cc`? Does highlighting work for you, when using my init.vim and scenario 1?

I came up with a Dockerfile to reproduce what I'm seeing, only to find out that semantic highlighting works fine for files I open from the command line, but it does *not* work for me for files opened from within the editor.

**Comment by jackguo380** [comment]
Hi @kmhofmann, thanks I was able to reproduce this one. I found it kind of strange that it was just LanguageClient integration being broken since I use it daily so I looked into it more. As it turns out this bug only appears on my appimage neovim (Neovim 0.5.0). I tried it on Neovim 0.4.3 and Vim 8.2 and this bug does not happen.

I tried debugging the why and found out that in Neovim 0.5.0 the autocmd that LanguageClient uses to notify plugins for whatever reason doesn't work when registered in a autoload function.
https://github.com/jackguo380/vim-lsp-cxx-highlight/blob/20f0f7c10f61368992f2e7b4edfc4ca54aa7d09a/autoload/lsp_cxx_hl/client/LanguageClient.vim#L9

I added a workaround for now. Let me know if this fixes it for you.
